include/pmacc/lockstep/ForEach.hpp

```diff
@@ -0,0 +1,226 @@
+/* Copyright 2017-2021 Rene Widera
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc/lockstep/Config.hpp"
+#include "pmacc/lockstep/Idx.hpp"
+#include "pmacc/lockstep/Variable.hpp"
+#include "pmacc/lockstep/Worker.hpp"
+#include "pmacc/types.hpp"
+
+namespace pmacc
+{
+    namespace lockstep
+    {
+        /** Execute a functor for the given index domain.
+         *
+         * Algorithm to execute a subsequent lockstep for each index of the configured domain.
+         * @attention There is no implicit synchronization between workers before or after the execution of ForEach
+         * performed.
+         *
+         * @tparam T_Config Configuration for the domain and execution strategy.
+         *                  T_Config must provide: domainSize, numCollIter, numWorkers, and simdSize at compile time.
+         */
+        template<typename T_Config>
+        struct ForEach
+            : T_Config
```

Comment:
I would prefer not template-parametrizing for any `T_Config` but having `Config<T_a, T_b, T_c>` since this is not supposed to work with any other class and would make it easier to follow. But no big deal.